### PR TITLE
Store password encrypted

### LINKED ISSUE #244: Secure mail account passwort in db
Passwort is stored cleartext in db, should be saved encrypted.

This should help:
https://github.com/owncloud/core/pull/10642

### PR DESCRIPTION
**Pull request by DeepDiver1975**
fixes #244

**PLEASE TRUNCATE YOUR DATABASE ENTRIES**

### CODE REVIEW

**Comment by jancborchardt** [comment]
Truncate as in resetting how it’s described in the README? https://github.com/owncloud/mail#how-do-i-reset-the-mail-app

@LukasReschke please review this security-related PR. ;)

**Comment by jancborchardt** [comment]
Also cc @Henni @4lexwe because we talked about this at the meetup. Review if you like. :)

**Comment by jancborchardt** [comment]
Also cc @Henni @irgendwie because we talked about this at the meetup. Review if you like. :)

**Comment by irgendwie** [comment]
@DeepDiver1975
Checked out fix-244 and truncated the database.
After activation of the app, the webpage shows:

*Fatal error: Cannot use object of type Closure as array in /opt/lampp/htdocs/oc/lib/private/navigationmanager.php on line 24*

And I can find the following lines in the log:

*{"reqId":"952b2bdb779383483c155c970015d705","remoteAddr":"::1","app":"PHP","message":"Argument 1 passed to OC\\NavigationManager::add() must be of the type array, object given, called in \/opt\/lampp\/htdocs\/oc\/apps\/mail\/appinfo\/app.php on line 26 and defined at \/opt\/lampp\/htdocs\/oc\/lib\/private\/navigationmanager.php#23","level":0,"time":"2015-04-20T12:22:54+00:00","method":"GET","url":"\/oc\/"}
{"reqId":"952b2bdb779383483c155c970015d705","remoteAddr":"::1","app":"PHP","message":"Cannot use object of type Closure as array at \/opt\/lampp\/htdocs\/oc\/lib\/private\/navigationmanager.php#24","level":3,"time":"2015-04-20T12:22:54+00:00","method":"GET","url":"\/oc\/"}*

**Comment by Bonbadil** [comment]
@irgendwie 
I'm using master (2d94d2e944) and I already have this error. It may be not related to this PR?
I'm trying to revert to get back my cloud.

**Comment by jancborchardt** [comment]
Sounds like this is the fix for it https://github.com/owncloud/mail/pull/512

Can you try that out @Bonbadil @irgendwie?

**Comment by jancborchardt** [comment]
@Bonbadil should be fixed on master by merging https://github.com/owncloud/mail/pull/512
Btw, you’re also welcome in our #owncloud-mail IRC channel :)

**Comment by Bonbadil** [comment]
@jancborchardt Thank you very much!

**Comment by irgendwie** [comment]
Ok @jancborchardt , App works now - but account creation fails this time:

*
...{"reqId":"9642b81a87f44d75c526631347444bb1","remoteAddr":"::1","app":"mail","message":"Creating account failed: Authenticated ciphertext could not be decoded.","level":3,"time":"2015-04-20T13:35:34+00:00","method":"POST","url":"\/oc\/index.php\/apps\/mail\/accounts"}...*

**Comment by Gomez** [comment]
Same here (with other errors): 
```
{"reqId":"kdZiJ0E4t\/xKp\/hhxHqm","remoteAddr":"127.0.0.1","app":"mail","message":"Creating account failed: Authenticated ciphertext could not be decoded.","level":3,"time":"2015-04-22T07:54:15+00:00"}
{"reqId":"kdZiJ0E4t\/xKp\/hhxHqm","remoteAddr":"127.0.0.1","app":"PHP","message":"Undefined variable: l10n at \/home\/gomez\/public_html\/mail_dev\/core\/apps\/mail\/lib\/controller\/accountscontroller.php#191","level":3,"time":"2015-04-22T07:54:15+00:00"}
{"reqId":"kdZiJ0E4t\/xKp\/hhxHqm","remoteAddr":"127.0.0.1","app":"PHP","message":"Cannot access empty property at \/home\/gomez\/public_html\/mail_dev\/core\/apps\/mail\/lib\/controller\/accountscontroller.php#191","level":3,"time":"2015-04-22T07:54:15+00:00"}
```

**Comment by DeepDiver1975** [comment]
I will look into the account creation ....

**Comment by DeepDiver1975** [comment]
Please retest - THX

**Comment by jancborchardt** [comment]
@Bonbadil @irgendwie @Gomez can you test this again? :)

**Comment by DeepDiver1975** [comment]
grrrr - travis fails .... will have a look ....

File: lib/account.php [hunk 1]
```diff
@@ -69,18 +72,19 @@ public function getImapConnection() {
 			$host = $this->account->getInboundHost();
 			$user = $this->account->getInboundUser();
 			$password = $this->account->getInboundPassword();
+			$password = $this->crypto->decrypt($password);
```
Comment (irgendwie) [inline lib/account.php]: Problem still exists - at this point the password gets decrypted although at this moment the password is unencrypted in the variable.

Comment (DeepDiver1975) [inline lib/account.php]: > Problem still exists - at this point the password gets decrypted although at this moment the password is unencrypted in the variable.

which is not a case we cover - the password has to be encrypted as taken from the db.

Comment (DeepDiver1975) [inline lib/account.php]: hmmm - travis fails as well


````
There were 2 errors:

1) OCA\Mail\Tests\Imap\AccountTest

exception 'Exception' with message 'Authenticated ciphertext could not be decoded.' in /home/travis/build/owncloud/core/lib/private/security/crypto.php:113

Stack trace:

#0 /home/travis/build/owncloud/core/apps/mail/lib/account.php(75): OC\Security\Crypto->decrypt('xxxxx')

#1 /home/travis/build/owncloud/core/apps/mail/tests/imap/abstracttest.php(50): OCA\Mail\Account->getImapConnection()

#2 [internal function]: OCA\Mail\Tests\Imap\AbstractTest::setUpBeforeClass()

#3 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestSuite.php(673): call_user_func(Array)

#4 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestSuite.php(716): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))

#5 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/TestRunner.php(398): PHPUnit_Framework_TestSuite->run(Object(PHPUnit_Framework_TestResult))

#6 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/Command.php(152): PHPUnit_TextUI_TestRunner->doRun(Object(PHPUnit_Framework_TestSuite), Array)

#7 phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/Command.php(104): PHPUnit_TextUI_Command->run(Array, true)

#8 /home/travis/.phpenv/versions/5.4.37/bin/phpunit(722): PHPUnit_TextUI_Command::main()

Comment (irgendwie) [inline lib/account.php]: "as taken from the db" - you also call the function **before** you save the account in the database (at account creation - so only **once**), that's why it fails at this point. Later calls use a password from the database, which is encrypted.

So after the account is in the database (after the imap and smtp tests), your code works as expected.

Comment (DeepDiver1975) [inline lib/account.php]: during setup the password gets encrypted as well: https://github.com/owncloud/mail/blob/fix-244/lib/service/autoconfig.php#L384-L388

**Comment by DeepDiver1975** [comment]
I just retested the setup mechanism - this is working from my pov.

**Comment by jancborchardt** [comment]
:+1: Also works for me now. The Firefox problem is a different issue: https://github.com/owncloud/mail/issues/524

Can we have another test please @wurstchristoph @irgendwie @Henni @Gomez?

**Comment by Gomez** [comment]
Configure with "Manual configuration" throws the error again for me:
```
{"reqId":"nGYh+e\/RhwSJwDFu4IZC","remoteAddr":"127.0.0.1","app":"mail","message":"Creating account failed: Authenticated ciphertext could not be decoded.","level":3,"time":"2015-04-24T10:01:44+00:00"}
```
Anyone tested with manual, too?

**Comment by DeepDiver1975** [comment]
anyone tested with manual too

**Comment by Xenopathic** [comment]
It might be a good idea to move the crypto stuff out of `autoconfig.php` into `accountcontroller.php`, so encryption for both manual accounts and auto accounts are in the same place. A simple `set*Password($this->crypto->encrypt(...->get*Password()));` should do the trick, right before saving to the DB?

**Comment by Xenopathic** [comment]
```
Missing argument 10 for OCA\Mail\Controller\AccountsController::__construct(), called in /srv/http/owncloud/apps/mail/appinfo/application.php on line 62 and defined at /srv/http/owncloud/apps/mail/lib/controller/accountscontroller.php#90
```

**Comment by DeepDiver1975** [comment]
@Xenopathic @Gomez  mind retesting

**Comment by Xenopathic** [comment]
```
Creating account failed: Authenticated ciphertext could not be decoded.
```

Manual set up (note, I hit the undefined variable bug I fixed in #534)

**Comment by jancborchardt** [comment]
@DeepDiver1975 be like

**Comment by Gomez** [comment]
works! Nice work. Cool commit history...

**Comment by Xenopathic** [comment]
Hooray! Tested with manual and auto accounts, works great! Can confirm that passwords are encrypted in the DB. Still unsure about the two places that passwords get encrypted during setup, but otherwise this is OK :+1:

**Comment by jancborchardt** [comment]
Ok, merged! Everyone, remember you need to reset your installation cause this changes the database:
https://github.com/owncloud/mail#resetting-the-app

cc @zinks- @wurstchristoph @LukasReschke @Henni @irgendwie @PoPoutdoor @Bonbadil

**Pull request merged**